### DISCUSSION TITLE
collaboration and contribution

### DISCUSSION THREAD

**Discussion by k41p10u** [PaulRBerg/hardhat-template#35] [General]
Hey @paulrberg 

A few months ago I started some Solidity development and in the beginning I found your template code to be tremendously helpful.

In the meantime I have added some extra features into my own project in terms of:

- using the `OpenZeppelin` contracts as a base
- more advanced test cases that:
  - asserts when errors are thrown,
  - asserts that is is the only the owner that is performing the action
  - asserts that an event is emitted from the contract method
  - etc.

which I think may be helpful to folks starting off on the Solidity development and who would like to include these more advanced features as well, but may struggle a bit to get it setup to work as such, as I had to tweak some of the src code quite a bit to work with these setups.

So this is a shoutout for a discussion perhaps if you might be interested in such a contribution to this `template project`.

_k41p10u_

**Comment by PaulRBerg** [comment]
Hey @k41p10u, glad that you found this template helpful!

As with anything in software engineering, there are trade-offs to consider when adding more complexity. One of the perks of my Solidity template is its simplicity, and I would like to keep that as such. I feel like stuffing in the OpenZeppelin contracts would be annoying for the template users who are not interested in depending on a third-party library.

However, I do agree that having more test cases would be helpful.
